### DISCUSSION TITLE
SmartCodable 如何支持继承的解析？

### DISCUSSION THREAD

**Discussion by intsig171** [intsig171/SmartCodable#34] [Ideas]
面对使用者的需求，我们不得不考虑如何处理继承关系。 截止到现在，基于Codable的解析方案，只有CodableWrapper支持了继承的能力。 欢迎提供有用的线索。

**Comment by faimin** [comment]
[keyedCodable](https://github.com/dgrzeszczak/KeyedCodable) 对继承的场景采用的是组合的方案，而非继承。[KeyedCodable #8](https://github.com/dgrzeszczak/KeyedCodable/issues/8)

**Comment by intsig171** [comment]
![image](https://github.com/intsig171/SmartCodable/assets/87351449/6fea7b28-87c8-46a2-900e-208364212d4c)
这种内部的解析组合很好实现。 如图所说。   可是使用上还是跟继承有区别。

**Comment by intsig171** [comment]
你理解KeyedCodable中的数组的举例是什么意思么？ 
![image](https://github.com/intsig171/SmartCodable/assets/87351449/f60e4598-8acd-479a-8ac5-bdc904e38fa3)
这本来就是一种嵌套结构，可以直接解析成功吧？ 看上去似乎不需要使用Flat。

**Reply by faimin** [reply]
不知道，项目中并没有真正使用`KeyedCodable`，只是在很早之前想找一个方便的、对于继承场景也不需要写很多映射的 `codable` 封装库发现的它，最终也没找到合适的（继承都避免不了写很多映射代码），但觉得`KeyedCodable`的思路不免为一个非常好的折中方案

**Comment by intsig171** [comment]
他的方案很好实现。我正在开发支持这个功能。 是不是用协议代替父类，让子类遵循该协议（自然会有父类的属性）达到继承的目的。 这样是不是更好一点？

**Reply by faimin** [reply]
你的思路挺好的，不过我个人觉得在使用体验上还是有些繁琐（需要把父类的属性在协议中抽离一遍），最终使用者对此种方案是否买账也很难说。个人建议可以暂时先提供一种实现简单的方案来满足对此有强诉求的用户，再慢慢的调研更完美的方案。

**Comment by intsig171** [comment]
最新的V4.0.4版本已经支持@Flat。
![d5d6b1db589df35b8b44ae8d705e9391](https://github.com/intsig171/SmartCodable/assets/87351449/2f7cee01-da4f-4bf5-9a1c-64876fb6c6b2)
我会继续调研最优的解决方案。

**Comment by faimin** [comment]
有位老哥用宏极大的简化了`Codable` 的继承处理，可以借鉴一下。 [CodableWrapper](https://github.com/winddpan/CodableWrapper)

**Comment by iAmMccc** [comment]
SmartCodable V5.0+ 已支持继承
```
class BaseModel: SmartCodable {
    var name: String = ""
    required init() { }
}

@SmartSubclass
public class StudentModel: BaseModel {
    var age: Int?
}
```
